### PR TITLE
Add a function to find missing cherry picks in different branches

### PR DESCRIPTION
**Pull request by juliagmt-google**
Context: https://github.com/pytorch/pytorch/issues/115469

### CODE REVIEW

**Comment by juliagmt-google** [comment]
command to run:
`python github_analyze.py --repo-path ~/pytorch --remote upstream --branch release/2.2  --target_branch release/2.3 --missing-in-release`

**Comment by atalman** [comment]
Couple of notes. I see it produces output like this:
```
1518578b54bc6881fd7eaff9bc2489b39132c35a ;  Iris Z <31293777+wz337@users.noreply.github.com> ;  2023-12-12 12:05:40
e57f089704ae9fa0f57f18c6122ce645457b183a ;  Nikita Shulga <2453524+malfet@users.noreply.github.com> ;  2023-12-08 19:33:13
44d11579dbe94b364bdd350bb94ff19c91731cb5 ;  Andrey Talman <atalman@fb.com> ;  2023-12-07 15:14:33
0863b4c354641ee60443a94b58562e3ad95feb8f ;  Jae Hoon (Antonio) Kim <17433012+antoniojkim@users.noreply.github.com> ;  2023-12-07 09:47:54
12bcfddce5773410d6e859e3c0feda9dc6c9b8d3 ;  Andrey Talman <atalman@fb.com> ;  2023-12-07 09:43:05
99718eda57e04dfdc6537325e7dcc962c625c3fd ;  Andrey Talman <atalman@fb.com> ;  2023-12-07 07:29:28
24397727a8500e2ec598a5ae3efb72d153da9b64 ;  Andrey Talman <atalman@fb.com> ;  2023-12-06 18:49:00
54aca571d6ded30bc5e6e19051bd460db6a631d4 ;  Andrey Talman <atalman@fb.com> ;  2023-12-06 15:30:25
```

This format is hard to work with, we are looking for similar output as in ``missing-in-branch `` :
``
PR or Issue link; Title; Status
``
This way issue or PR can be clicked and viewed right away.

```
python analytics/github_analyze.py --repo-path ~/local/pytorch --remote upstream --branch release/2.2 -
-milestone-id 43 --missing-in-branch 
https://github.com/pytorch/pytorch/issues/120999;[v2.2.2] Release Tracker;open
https://github.com/pytorch/pytorch/issues/120547;F.conv1d fails with `RuntimeError: cannot create std::vector larger than max_size()` after transpose of input tensor;closed
``

**Comment by juliagmt-google** [comment]
`python github_analyze.py --repo-path ~/pytorch --remote upstream --branch release/2.2  --target-branch release/2.3 --missing-in-release --date 2024-03-07`

It will output all the missing cherry picks (PR html, PR title, PR status) in `--branch release/2.2`after `--date 2024-03-07` for `--target-branch release/2.3`

Found 11 missing cherry picks below, could you please help me validate? 
-----------------------------------------
URL;Title;Status
https://github.com/pytorch/pytorch/pull/122207 ;  Pin protobuf to 3.20.2 on macOS ;  closed
https://github.com/pytorch/pytorch/pull/122120 ;  Fix MSVC 14.38 - VS 2022 Build ;  closed
https://github.com/pytorch/pytorch/pull/122122 ;  chore: add unit test to verify split_by_tags output_type ;  closed
https://github.com/pytorch/pytorch/pull/122121 ;  fix: set codegen in _SplitterBase partitioner ;  closed
https://github.com/pytorch/pytorch/pull/120518 ;  Fix missing permission in create release workflow ;  closed
https://github.com/pytorch/pytorch/pull/121934 ;  [Release only] Disable triton build workflows ;  closed
https://github.com/pytorch/pytorch/pull/121419 ;  [CPP] Update GCC minversion check to 9 or newer ;  closed
https://github.com/pytorch/pytorch/pull/121026 ;  Windows Dynamo Error Removal CI Check ;  closed
https://github.com/pytorch/pytorch/pull/121250 ;  [FSDP][StateDict] Allow FULL_STATE_DICT option for 2D (#120837) ;  closed
https://github.com/pytorch/pytorch/pull/121229 ;  Fix make triton command on release branch (#121169) ;  closed
https://github.com/pytorch/pytorch/pull/121232 ;  Don't use size on TensorVariable when doing out resize test ;  closed

**Comment by juliagmt-google** [comment]
Updated the logic and here is the new output for missing PRs in `release/2.3` after 2024-03-07 from cherry picks in `release/2.2` 

```
len(main_commits)=336
len(cherry_pick_commits)=76
https://github.com/pytorch/pytorch/issues/120999;[v2.2.2] Release Tracker;open
https://github.com/pytorch/pytorch/pull/120361;fix: set codegen in _SplitterBase partitioner;closed
```

**Comment by atalman** [comment]
HI @juliagmt-google looks like the output is not correct. This cherry-pick:
https://github.com/pytorch/pytorch/commit/15add24bf28477843a7e13d9deaa4beb39473900 is part of the release 2.3 . One way to check, it was merged in main before the date 2024-03-07

**Comment by juliagmt-google** [comment]
`python github_analyze.py --repo-path ~/pytorch --remote upstream --branch release/2.2  --minor-release release/2.3 --missing-in-release --date 2024-03-07 --milestone-id 43 --issue-num 121760`

len(main_commits)=336
len(prev_release_commits)=76
len(current_issue_comments)=13
issue_num: 121760, len(issue_comments)=13
URL;Title;Status
https://github.com/pytorch/pytorch/pull/120361; fix: set codegen in _SplitterBase partitioner;closed

File: analytics/github_analyze.py [hunk 1]
```diff
@@ -372,6 +373,28 @@ def get_commits_dict(x, y):
...
+    main_commits = get_commits_dict(minor_release, 'main')
+    prev_release_commits = get_commits_dict(orig_branch, branch)
+    current_issue_comments = get_issue_comments('pytorch', 'pytorch',issue_num) # issue comments for the release tracker as cherry picks
+    print(f"len(main_commits)={len(main_commits)}")
+    print(f"len(prev_release_commits)={len(prev_release_commits)}")
+    print(f"len(current_issue_comments)={len(current_issue_comments)}")
+    print(f"issue_num: {issue_num}, len(issue_comments)={len(current_issue_comments)}")
+    print("URL;Title;Status")
+
+    for issue in gh_get_milestone_issues('pytorch', 'pytorch', milestone_idx, IssueState.ALL):
```
Comment (atalman) [inline analytics/github_analyze.py:388]: Here we should not iterate over milestone. I think would be best to iterate over the previous release branch.

Something along the lines:
```
for commit in prev_release_commits.values():
        not_cherry_picked_in_current_issue = any(commit.pr_url not in issue_comment['body'] for issue_comment in current_issue_comments)
        for main_commit in main_commits.values():
            if main_commit.pr_url == commit.pr_url :
                mentioned_after_cut_off_date = cut_off_date < main_commit.commit_date
                if not_cherry_picked_in_current_issue and mentioned_after_cut_off_date:
                    print(f"{commit.commit_date} {commit.title} {commit.author} {commit.pr_url}")
                break
```


For this we would need to alter the parse_fuller_format:
```
def parse_fuller_format(lines: Union[str, List[str]]) -> GitCommit:
    """
    Expect commit message generated using `--format=fuller --date=unix` format, i.e.:
        commit <sha1>
        Author:     <author>
        AuthorDate: <author date>
        Commit:     <committer>
        CommitDate: <committer date>

        <title line>

        <full commit message>

    """
    if isinstance(lines, str):
        lines = lines.split("\n")
    # TODO: Handle merge commits correctly
    if len(lines) > 1 and lines[1].startswith("Merge:"):
        del lines[1]
    assert len(lines) > 7
    assert lines[0].startswith("commit")
    assert lines[1].startswith("Author: ")
    assert lines[2].startswith("AuthorDate: ")
    assert lines[3].startswith("Commit: ")
    assert lines[4].startswith("CommitDate: ")
    assert len(lines[5]) == 0

    prUrl = ""
    for line in lines:
        if "Pull Request resolved:" in line:
            prUrl = line.split("Pull Request resolved:")[1].strip()
            break

    return GitCommit(commit_hash=lines[0].split()[1].strip(),
                     author=lines[1].split(":", 1)[1].strip(),
                     author_date=datetime.fromtimestamp(int(lines[2].split(":", 1)[1].strip())),
                     commit_date=datetime.fromtimestamp(int(lines[4].split(":", 1)[1].strip())),
                     title=lines[6].strip(),
                     body="\n".join(lines[7:]),
                     pr_url=prUrl,
                     )
```

**Pull request merged**